framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -528,56 +589,135 @@ private AnnotationMirror createBooleanAnnotationMirror(List<Boolean> values) {
                 }
             }
         }
+
+        // Refine the values of the operands in the store based on the result of the comparison.
+
+        List<Number> thenLeftVals = new ArrayList<>();
+        List<Number> elseLeftVals = new ArrayList<>();
+
+        List<Number> thenRightVals = new ArrayList<>();
+        List<Number> elseRightVals = new ArrayList<>();
+
+        // Because resultValues contains the (boolean) results of each comparison (i.e. the product of the number
+        // of elements in each list), we have to iterate through parts of the list at a time in order to find all
+        // the results of comparisons involving a particular value in the left list.
+        for (int i = 0; i < lefts.size(); i++) {
+            for (int j = 0; j < rights.size(); j++) {
+                if (resultValues.get(j + i * rights.size())) {
+                    thenLeftVals.add(lefts.get(i));
+                    thenRightVals.add(rights.get(j));
+                } else {
+                    elseLeftVals.add(lefts.get(i));
+                    elseRightVals.add(rights.get(j));
+                }
+            }
+        }
+
+        CFStore thenStore = result.getThenStore();
+        CFStore elseStore = result.getElseStore();
+
+        createAnnotationFromResultsAndAddToStore(thenStore, thenLeftVals, leftNode);
+        createAnnotationFromResultsAndAddToStore(elseStore, elseLeftVals, leftNode);
+
+        createAnnotationFromResultsAndAddToStore(thenStore, thenRightVals, rightNode);
+        createAnnotationFromResultsAndAddToStore(elseStore, elseRightVals, rightNode);
+
         return resultValues;
     }
 
+    private void createAnnotationFromResultsAndAddToStore(
+            CFStore store, List<?> results, Node node) {
+        // If a zero-length list is passed, this returns bottom - which is very problematic.
+        AnnotationMirror anno =
+                atypefactory.createResultingAnnotation(
+                        node.getType(), results.size() == 0 ? null : results);
+        AnnotationMirror currentAnno =
+                atypefactory
+                        .getAnnotatedType(node.getTree())
+                        .getAnnotationInHierarchy(atypefactory.BOTTOMVAL);
+        Receiver rec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), node);
+        store.insertValue(
+                rec, atypefactory.getQualifierHierarchy().greatestLowerBound(anno, currentAnno));
```

Comment:
Am I correct that GLB will only return bottom if we have two inconsistent sets of values? For instance, if one side is `@IntVal({1,2})` and the other is `@IntVal({3})`? In those cases, I think we're doing the right thing: in dead code, the refined types of things should be bottom, since we don't want to issue alarms.